### ISSUE TITLE
Clarification on Trivial Splits

### ISSUE THREAD

**Issue by cowden** [dmlc/xgboost#2914]
I have a question regarding how xgboost splits out missing values.  The exact greedy and approximate tree methods tend to split out missing values by splitting at twice the maximum value on a feature.  Whereas the histogram method splits at the minimum value of a feature.  The factor of 2 seems consistent, so I don't think it has any dependence on the distribution of the feature.

Just for reference, I've been able to consistently see this kind of split in the latest release (v0.60). 

I've found some threads discussing the treatment of missing data [2],[3],[4], but I'm interesting in understanding why 2?

So my question is the following:
Is the decision to exclusively split away missing values at twice the maximum of a variable a consequence/artifact/feature of the split proposal algorithm described in [1], or is it a feature of the implementation?

If it's the latter case, can this parameter be tuned?

[1] https://arxiv.org/abs/1603.02754
[2] https://github.com/dmlc/xgboost/issues/21
[3] https://github.com/dmlc/xgboost/issues/1112
[4] https://github.com/dmlc/xgboost/issues/1842

**Comment by hcho3** [comment]
I've worked on the histogram implementation and here's my take on it:

1) The factor 2 is simply an implementation choice; there's nothing inherent in the algorithm that forces it. It's probably to give enough cushion to ensure all non-missing entries fall on one side of the split. The factor 2 is introduced by a choice in the approximate quantile sketcher:
```
/* Lines 95-101 of src/common/hist_util.cc */
    // push a value that is greater than anything
    if (a.size != 0) {
      bst_float cpt = a.data[a.size - 1].value;
      // this must be bigger than last value in a scale
      bst_float last = cpt + fabs(cpt);
      cut.push_back(last);
    }
```
So the factor 2 is currently not tune-able.

2) As for the choice between minimum and maximum values, we should look at how candidate splits are enumerated. In exact and approximate methods, the candidate splits are scanned from left to right, and then right to left:
```
/* Lines 630-637 of updater_colmaker.cc */
          if (param.need_forward_search(fmat.GetColDensity(fid), ind)) {
            this->EnumerateSplit(c.data, c.data + c.length, +1,
                                 fid, gpair, info, stemp[tid]);  // note +1
          }
          if (param.need_backward_search(fmat.GetColDensity(fid), ind)) {
            this->EnumerateSplit(c.data + c.length - 1, c.data - 1, -1,
                                 fid, gpair, info, stemp[tid]);  // note -1
          }
```
The left-to-right scan considers the maximum feature value, and the right-to-left considers the minimum feature value. This is to take account of the "trivial split."

The histogram method does the same thing in different order: the candidates are scanned from right to left, and then from left to right:
```
/* Lines 482-485 of updater_colmaker.cc */
        this->EnumerateSplit(-1, gmat, hist[nid], snode[nid], constraints_[nid], info,
          &best_split_tloc_[tid], fid);  // note -1
        this->EnumerateSplit(+1, gmat, hist[nid], snode[nid], constraints_[nid], info,
          &best_split_tloc_[tid], fid);  // note +1
```
I wrote this piece of code 8 months ago. I recall responding to a tricky bug that was activated only when the scanning direction was right-to-left; that's probably why the first scan is set to right-to-left in the histogram method.

In short, the choice of minimum/maximum values is again an implementation detail.

So here's a question for you: does the choice between minimum and maximum splits matter for your use case? Would it useful to be able to specify it?

**Comment by cowden** [comment]
Thanks so much for the clarification.  It makes perfect sense now.

In my case I have many non-negative features.  So, splitting trivial splits from the minimum comes with the guarantee that future non-missing data will not fall into that split, as opposed to the small chance of that happening on the right side.  To have the same functional result, why not have the certainty.  

I'm trying to think if tuning the factor of two, or rather setting it to a significance level which is tuned, makes sense.  It's not obvious to me how the probability holds up locally in a tree, but perhaps it's a good enough approximation of a rare event.

**Comment by hcho3** [comment]
@cowden One easy way out is simply allow `inf` and `-inf` for thresholds, as LightGBM currently does.

**Comment by cowden** [comment]
Good idea.  I'll look into it.

**Comment by hcho3** [comment]
Consolidating to #3439. This issue should be re-opened when someone decides to actively work to clarify trivial splits.

**Comment by trivialfis** [comment]
> someone decides to actively work to clarify trivial splits.

Me.

**Comment by hcho3** [comment]
Note: Using `inf` and `-inf` for trivial splits will let us remove `min_fvalue` and `min_vals_` fields.
